Neos.Neos/Classes/Fusion/MenuItemsImplementation.php

```diff
@@ -312,14 +318,32 @@ protected function traverseUpUntilCondition(Node $node, \Closure $callback): voi
         } while ($shouldContinueTraversal !== false && $node !== null);
     }
 
-    protected function buildUri(Node $node): string
+    public function getCurrentNodeAncestors(): Nodes
+    {
+        if ($this->currentNodeAncestors instanceof Nodes) {
+            return $this->currentNodeAncestors;
+        }
+        $subgraph = $this->contentRepositoryRegistry->subgraphForNode($this->currentNode);
+        $this->currentNodeAncestors = $subgraph->findAncestorNodes(
```

Comment:
This will potentially load thousands and thousands of nodes into memory.. I don't think that we can pre-cache this

Reply:
it should be super fine as this is opt in behavior.

Reply:
Really?
Do you think an integrator is aware that we load all pages into memory if they set this flag to true to get the Neos < 9 behavior ??

Reply:
No, but my argument would be: If you want speed you know that you only want one cache entry https://mind-the-seb.de/blog/globally-cached-fragments-in-neos-cms and then you know that you dont want to have this calculate item flag -> speeeeeed.

Reply:
I think we are talking cross purpose.
I agree of having that flag for exactly that reasons.
I don't agree of a default implementation that loads all documents into memory – especially because you will only see a crazy increase of RAM usage and no direct indication to why that is

Reply:
Not sure i understand that correctly. As far as i see `findAncestorNodes` leads to a single (internally recursive) db-query that will return only the nodes up from the given node->id. 1000s of nodes would require a structure 1000 levels deep that will cause other issued aswell. 

Is that somehow correct or am i on the woodway.

Reply:
You're right.. I confused myself and mixed up descendants and ancestors for a moment #coffee 😶‍🌫️

Reply:
Sorry for the confusion. Loading the _ancestors_ once per menu should be fine.
But I would still suggest to only store the node *ids*  instead of the instances with all their properties etc.

We can use `NodeAggregateIds` VO for that (and should extend that with some `contains(NodeAggregateId $nodeAggregateId): bool` to simplify the check below)

Reply:
Adjusted as requested